[HLSL] set alwaysinline on HLSL functions

pow2clk opened the pull request:
HLSL inlines all its functions by default. This uses the alwaysinline attribute to force that in the corresponding pass for user functions by default and overrides the default noinline of some implicit functions. This makes an instance of explicit inlining for buffer subscripts unnecessary.

Adds tests for function and constructor inlining and augments some existing tests to verify correct inlining of implicitly created functions as well.

incidentally restore RUN line that I believe was mistakenly removed as part of #88918

fixes #89282

Commit:
[HLSL] set alwaysinline on HLSL functions

HLSL inlines all its functions by default. This uses the alwaysinline
attribute to force that in the corresponding pass for user functions
by default and overrides the default noinline of some implicit functions.
This makes an instance of explicit inlining for buffer subscripts unnecessary.

Adds tests for function and constructor inlining and augments some existing
tests to verify correct inlining of implicitly created functions as well.

incidentally restore RUN line that I believe was mistakenly removed as part of #88918

fixes #89282

Commit:
update RWBuffer-AST test for later alwaysinline marking

Previously, the alwaysinline attribute was set for the RWBuffer
subscript operator as soon as it was created. Since they are all
in a common place now, that setting was redundant, but it does mean
that it won't show up in the AST

llvm-beanz commented on clang/lib/CodeGen/CGHLSLRuntime.cpp:
```diff
@@ -414,9 +414,20 @@ void CGHLSLRuntime::emitEntryFunction(const FunctionDecl *FD,
+  } else { // "implicit" autogenerated functions with no declaration
+    // Implicit functions might get marked as noinline by default
+    // but we override that for HLSL
+    Fn->removeFnAttr(Attribute::NoInline);
```
Which auto-generated functions are getting noinline attached to them? Do you know where that is happening?

llvm-beanz replied:
It seems to me like it might make more sense to change that condition to not do that for HLSL rather than to modify the IR after we generate it incorrectly.

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -290,8 +290,6 @@ struct BuiltinTypeDeclBuilder {
     MethodDecl->setLexicalDeclContext(Record);
     MethodDecl->setAccess(AccessSpecifier::AS_public);
-    MethodDecl->addAttr(AlwaysInlineAttr::CreateImplicit(
-        AST, SourceRange(), AlwaysInlineAttr::CXX11_clang_always_inline));
```
Why was it necessary to remove this? I think we want this because we'll want these calls inlined even in separate compilation cases.

llvm-beanz replied:
We don't really have a workflow in Clang implemented yet for separate compilation. The analog in DXC would be something like `-T lib_6_x -default-linkage=external`. In that case we would still want these functions to be inlined, and we can accomplish that with this attribute.

Commit:
Respond to feedback

Restore additional alwaysinline setting for RWBuffer subscripts

Add HLSL check to prevent marking autogenerated functions as noinline

pow3clk reviewed (commented):
On my laptop account. I assure you it's who you think it is.

Commit:
Default to safer behavior respecting noinline

For whatever reason a function might have noinline, if we try to set
alwaysinline, we get an incompatible attribute error. This avoids
that in any case.

Commit:
clang-format

damyanp commented on clang/lib/CodeGen/CodeGenFunction.cpp:
```diff
@@ -1239,9 +1239,9 @@ void CodeGenFunction::StartFunction(GlobalDecl GD, QualType RetTy,
 
-  if (FD && getLangOpts().HLSL) {
+  if (getLangOpts().HLSL) {
     // Handle emitting HLSL entry functions.
-    if (FD->hasAttr<HLSLShaderAttr>()) {
+    if (FD && FD->hasAttr<HLSLShaderAttr>()) {
       CGM.getHLSLRuntime().emitEntryFunction(FD, Fn);
     }
     CGM.getHLSLRuntime().setHLSLFunctionAttributes(FD, Fn);
```
Maybe not for this change, but a bit of me is wondering if the two CGHLSLRuntime functions emitEntryFunction and setHLSLFunctionAttributes should be somewhat combined so that we end up with something like:

```c++
if (getLangOpts().HLSL)
    CGM.getHLSLRuntime().emitFunctionProlog(FD, Fn);
```

Similar to how this is encapsulated for OpenMP above.  That way we get more of the HLSL-specific stuff in CGHLSLRuntime.cpp.

pow2clk replied:
Yeah, the `emitFunctionProlog` form is how it looked in an earlier version of a change Helena made. I think it was #102275 which introduced `setHLSLFucntionAttributes`. I decided to follow her lead, but I don't know the reasoning behind it. 

This becomes moot with the new approach I'm taking.

damyanp commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2474,7 +2474,9 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
     // disabled, mark the function as noinline.
+    // HLSL functions must be always inlined
     if (!F->hasFnAttribute(llvm::Attribute::AlwaysInline) &&
+        !getLangOpts().HLSL &&
```
A bit lower down there's some similar checks that seem to involve `AlwaysInline` and `OnlyAlwaysInlining` (around line 2536).  Does this also need to be updated for HLSL mode?

pow2clk replied:
I investigated this in some depth. The situation is that the autogenerated initialization and destruction functions get their attributes set through this call and others before their bodies are generated, which calls `setHLSLFunctionAttributes` while user functions generate their bodies and make the hlsl attributes call before this one. This seems error-prone to me, but the code has been around long enough that I'm not ready to go changing it even though some initial experiments showed no regressions. 

In the meantime, since so much inline attribute logic is in this function, I've concluded that the best solution is to add our inline logic here and not in `setHLSLFunctionAttributes` at all. That way it doesn't matter as much where in the IR function creation process it takes place and any consequences can be accounted for along with the existing logic. That's what my latest commit does.

Commit:
Revise approach entirely

Move all logic into SetLLVMFunctionAttributesForDefinition where
most inlining attribute calculations are done. The way autogenerated
init functions and user functions were created meant that this was hit
at different times for each relative to setHLSLFunctionAttributes.
Instead of trying to resolve that fundamental issue or work around it,
better to keep all the setting in the same place so it doesn't matter
where it is called relative to function body creation.

This required making exceptions for entry functions as they will get
attributes from this code as well and we want it to not be always
inlined and receive the relevant optimization attributes that allows.

pow2clk commented on clang/test/CodeGenHLSL/inline-functions.hlsl:
```diff
@@ -0,0 +1,114 @@
+
+// Verify that all functions have the alwaysinline attribute
+// CHECK: Function Attrs: alwaysinline
+// CHECK: define void @"?swap@@YAXY0GE@III@Z"(ptr noundef byval([100 x i32]) align 4 %Buf, i32 noundef %ix1, i32 noundef %ix2) [[IntAttr:\#[0-9]+]]
```
The reason why some of these won't be removed is because of the `-disable-llvm-passes` flag which skips optimizations that remove trivially dead functions.

This comment did make me realize there is a bug in waiting though. We don't yet remove trivially dead functions like this at all because the internal linkage marking takes place after any pass that would remove them. I have a tentative fix that depends on this change here #106146. Once this is in, I'll focus on that. Once that's in, these won't show up in the INLINE case, so I moved them to be NOINLINE exclusive.

pow2clk reviewed (commented):
Thanks Damyan!

Commit:
clang-format again

llvm-beanz commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2471,11 +2471,14 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
 
   if (!D) {
+    // HLSL functions must always be inlined
+    if (getLangOpts().HLSL && !F->hasFnAttribute("hlsl.shader"))
```
What if we put `noinline` on the entry function? Does that simplify this logic? Since we'll also at some point need to support the `noinline` keyword we'll need to be checking that attribute somewhere.

pow2clk replied:
In earlier discussion, we resolved to make this the default and not worry about noinline as of yet seeing as there are a lot of issues to resolve before that can work properly. In a previous version, I was forced to check for noinline as it was a sign that a function had passed through this function, but my preference is to not respect it until we've worked out the issues that prevent it from working right. 

 It wouldn't simplify the logic since the entry function doesn't get marked inline until it is passed into this function.

pow2clk replied:
Just an update. As a result of the waning minutes of a design discussion, we resolved to produce a warning when users apply noinline and explicitly set noinline for the outermost entry function which would allow checking for that in place of hlsl.shader here, which might allow graceful resolution of any other functions that happen to get it applied here which we could have an assert for.

llvm-beanz commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2471,11 +2471,14 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
     B.addAttribute(llvm::Attribute::StackProtectReq);
 
   if (!D) {
+    // HLSL functions must always be inlined
```
nit: we do try have comments be complete sentences and puctuated.
```suggestion
    // HLSL functions must always be inlined.
```

llvm-beanz commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2502,9 +2505,12 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
-  // Add optnone, but do so only if the function isn't always_inline.
-  if ((ShouldAddOptNone || D->hasAttr<OptimizeNoneAttr>()) &&
-      !F->hasFnAttribute(llvm::Attribute::AlwaysInline)) {
+  // alwaysinline all HLSL functions save entry points
```
nit: 
```suggestion
  // Add alwaysinline on all HLSL functions except entry points.
```

Commit:
reformat comments

damyanp reviewed (approved):
LGTM, but would be good to get a review from someone more knowledgeable in the area than I am.

Commit:
noinline adjustments: warning, entry function

Set noinline explicitly on the outermost entry function and use that
attribute to decide to apply alwaysinline.
Generate a warning and ignore user-specified noinline

Incidentally tidy up some RUN lines

Commit:
clang-format (again)

Commit:
Add test for new warning I left out of last commit

llvm-beanz commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -2502,9 +2505,16 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
+    if (D->hasAttr<NoInlineAttr>())
+      getDiags().Report(D->getLocation(),
+                        diag::warn_unsupported_attribute_ignored)
+          << "noinline" << "HLSL";
```
Why are we warning on this? Does the Clang implementation not work?

If we're going to warn on this, it needs to be in Sema not CodeGen!

llvm-beanz commented on clang/test/SemaHLSL/noinline.hlsl:
```diff
@@ -0,0 +1,13 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.0-compute -emit-llvm-only %s -verify
```
We should not need to emit IR to test a warning. This should work with `-fsyntax-only` or without the `-emit-llvm-only` flag.

Commit:
remove warning and respect noinline

nico commented:
Looks like the _revert_ broke check-clang: http://45.33.8.238/linux/148045/step_6.txt

Please take a look, and uh, mark that test as UNSUPPORTED if it takes a while to fix, I suppose.

nico commented:
@thurstond ^

thurstond commented:
@nico Sorry, my bad. I didn't notice the test was also updated as a companion to this change (https://github.com/llvm/llvm-project/pull/109023). I'll revert the test as well.

thurstond commented:
Oh, actually, I just realized the test update was the fix-forward. So the correct fix would be to reland both the change and the test update. Sorry!

thurstond commented:
I've reverted both of my reverts

pow2clk commented:
Sorry I didn't mention here that the fix was in. It looks like everything is clean now?

thurstond commented:
> Sorry I didn't mention here that the fix was in. It looks like everything is clean now?

Yes, both your changes are back in and I anticipate the buildbots will be green again. Sorry for the noise!
